File: modulo_6_cmms/internal/infrastructure/nats/publisher.go
package nats

import (
    "encoding/json"
    "log"
    "time"

    "github.com/nats-io/nats.go"
)

type EventPublisher struct {
    conn *nats.Conn
}

type MaintenanceEvent struct {
    EventType   string                 `json:"event_type"`
    Version     string                 `json:"version"`
    Timestamp   string                 `json:"timestamp"`
    Payload     map[string]interface{} `json:"payload"`
}

func NewEventPublisher(url string) (*EventPublisher, error) {
    nc, err := nats.Connect(url)
    if err != nil {
        return nil, err
    }
    log.Println("✅ Módulo 6: Conectado a NATS")
    return &EventPublisher{conn: nc}, nil
}

func (p *EventPublisher) PublishOrderCreated(payload map[string]interface{}) error {
    event := MaintenanceEvent{
        EventType: "MAINTENANCE_ORDER_CREATED",
        Version:   "1.0.0",
        Timestamp: time.Now().UTC().Format(time.RFC3339),
        Payload:   payload,
    }

    data, err := json.Marshal(event)
    if err != nil {
        return err
    }

    log.Printf("📤 Publicando evento: %s", event.EventType)
    return p.conn.Publish("maintenance.order.created", data)
}

func (p *EventPublisher) PublishOrderUpdated(payload map[string]interface{}) error {
    event := MaintenanceEvent{
        EventType: "MAINTENANCE_ORDER_UPDATED",
        Version:   "1.0.0",
        Timestamp: time.Now().UTC().Format(time.RFC3339),
        Payload:   payload,
    }

    data, err := json.Marshal(event)
    if err != nil {
        return err
    }

    log.Printf("📤 Publicando evento: %s", event.EventType)
    return p.conn.Publish("maintenance.order.updated", data)
}

func (p *EventPublisher) Close() {
    if p.conn != nil {
        p.conn.Close()
        log.Println("🛑 Módulo 6: Desconectado de NATS")
    }
}

File: modulo_10_hr_conductores/src/infrastructure/nats/publisher.ts
import { connect, StringCodec, NatsConnection } from 'nats';

export class NatsPublisher {
  private connection: NatsConnection | null = null;
  private sc = StringCodec();

  async connect(url: string): Promise<void> {
    try {
      this.connection = await connect({ servers: url });
      console.log('✅ Módulo 10: Conectado a NATS');
    } catch (error) {
      console.error('❌ Error conectando a NATS:', error);
    }
  }

  async publishDriverCreated(driver: any): Promise<void> {
    if (!this.connection) return;
    const event = {
      event_type: 'DRIVER_CREATED',
      version: '1.0.0',
      timestamp: new Date().toISOString(),
      payload: driver,
    };
    const data = this.sc.encode(JSON.stringify(event));
    this.connection.publish('driver.created', data);
    console.log(`📤 Evento publicado: DRIVER_CREATED - ${driver.id}`);
  }

  async publishDriverUpdated(driver: any): Promise<void> {
    if (!this.connection) return;
    const event = {
      event_type: 'DRIVER_UPDATED',
      version: '1.0.0',
      timestamp: new Date().toISOString(),
      payload: driver,
    };
    const data = this.sc.encode(JSON.stringify(event));
    this.connection.publish('driver.updated', data);
    console.log(`📤 Evento publicado: DRIVER_UPDATED - ${driver.id}`);
  }

  async close(): Promise<void> {
    if (this.connection) {
      await this.connection.close();
      console.log('🛑 Módulo 10: Desconectado de NATS');
    }
  }
}
